tests/unit/pubsubshard.tcl

```diff
@@ -1,80 +1,4 @@
 start_server {tags {"pubsubshard external:skip"}} {
-    proc __consume_ssubscribe_messages {client type channels} {
-        set numsub -1
-        set counts {}
-
-        for {set i [llength $channels]} {$i > 0} {incr i -1} {
-            set msg [$client read]
-            assert_equal $type [lindex $msg 0]
-
-            # when receiving subscribe messages the channels names
-            # are ordered. when receiving unsubscribe messages
-            # they are unordered
-            set idx [lsearch -exact $channels [lindex $msg 1]]
-            if {[string match "sunsubscribe" $type]} {
-                assert {$idx >= 0}
-            } else {
-                assert {$idx == 0}
-            }
-            set channels [lreplace $channels $idx $idx]
-
-            # aggregate the subscription count to return to the caller
-            lappend counts [lindex $msg 2]
-        }
-
-        # we should have received messages for channels
-        assert {[llength $channels] == 0}
-        return $counts
-    }
-
-    proc __consume_subscribe_messages {client type channels} {
-        set numsub -1
-        set counts {}
-
-        for {set i [llength $channels]} {$i > 0} {incr i -1} {
-            set msg [$client read]
-            assert_equal $type [lindex $msg 0]
-
-            # when receiving subscribe messages the channels names
-            # are ordered. when receiving unsubscribe messages
-            # they are unordered
-            set idx [lsearch -exact $channels [lindex $msg 1]]
-            if {[string match "unsubscribe" $type]} {
-                assert {$idx >= 0}
-            } else {
-                assert {$idx == 0}
-            }
-            set channels [lreplace $channels $idx $idx]
-
-            # aggregate the subscription count to return to the caller
-            lappend counts [lindex $msg 2]
-        }
-
-        # we should have received messages for channels
-        assert {[llength $channels] == 0}
-        return $counts
-    }
-
-    proc ssubscribe {client channels} {
-        $client ssubscribe {*}$channels
-        __consume_ssubscribe_messages $client ssubscribe $channels
-    }
-
-    proc subscribe {client channels} {
-        $client subscribe {*}$channels
-        __consume_subscribe_messages $client subscribe $channels
-    }
-
-    proc sunsubscribe {client {channels {}}} {
-        $client sunsubscribe {*}$channels
-        __consume_subscribe_messages $client sunsubscribe $channels
-    }
-
-    proc unsubscribe {client {channels {}}} {
-        $client unsubscribe {*}$channels
-        __consume_subscribe_messages $client unsubscribe $channels
-    }
```

Comment:
@hpatro any reason this code is (nearly) cloned of the one in util.tcl?
i.e. the one in util.tcl doesn't do an explicit `string match "sunsubscribe"` and `string match "unsubscribe"`, but instead allows both.

p.s. i noticed many of the tests here don't actually read the message they published from the subscribed connection.
is that because much of the testing of this feature is done in the cluster tests, so we allowed the tests here to be a bit lean?

Reply:
My bad, it got duplicated. util seems to be the right place for it.

I think I've covered most of the basic test cases here for a non cluster setup.  The place where message aren't read are to verify some other scenarios. Is there something you want more coverage on ? please let me know.

In cluster mode tests definitely there are more scenarios for sharded pubsub.